### PR TITLE
Add review guidelines to the default PR template

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

In the Zebra development process, we haven't made explicit decisions about:
- what should go in a pull request
- what to do about follow-up tasks
- what it means for code, documentation, and tests to be complete
- how we make sure the PR matches any design RFCs
- what things a reviewer should check

So we've been operating off conventions and assumptions we've learned from previous work.
Usually that works. But sometimes our assumptions clash, and we experience conflicts or confusion.

## Solution

This change adds reviewer guidelines, a follow-up section, and design RFC questions to the default GitHub pull request template.

In reviews, we want to focus on this overall goal, so it appears first, in bold:
**Does this pull request improve Zebra?**

The rest of the change is a list of useful things to check, split up into categories. Each category starts with a key question in bold. These checklists are based on my experience writing and reviewing PRs - please feel free to suggest changes.

I tried to keep each checklist to 3-7 items, so that they are readable.
(I have left the part of the checklist in this PR, so you can see how it is rendered.)

## Related Issues

- #208 - Add contributor guidelines on testing
- [RFC Template: Test Plan](https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0000-template.md#module-structure)

## Review Guidelines (for Reviewer)

**Does this pull request improve Zebra?**

- [ ] Pull Request
  - The pull request is complete
  - The PR contains changes to related code (unrelated changes can be split into another PR)
  - Any follow-up tasks are listed in a GitHub issue

- [ ] Documentation
  - Docs summarise how the template should be used
  - Docs reference specifications or design documents

## Zebra Team Approval

Since this is a process change, let's make sure the whole team has reviewed it:

- [ ] @dconnolly 
- [ ] @hdevalence 
- [ ] @oxarbitrage 
- [ ] @teor2345 
- [ ] @yaahc

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -26,3 +25,44 @@ the code change.
...
+This checklist can help you do your review.
+Add or skip tasks as needed.
+-->
+
+**Does this pull request improve Zebra?**
+
+- [ ] Pull Request
+  - The pull request is complete: code, documentation, and tests
+  - The PR contains changes to related code (unrelated changes can be split into another PR)
+  - Any follow-up tasks are listed in a GitHub issue
```
Comment (teor2345) [inline .github/pull_request_template.md:58]:
```suggestion
  - The PR does everything listed in the tickets that it closes
  - Any follow-up tasks are listed in a GitHub issue
```

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -26,3 +25,44 @@ the code change.
...
+
+- [ ] Documentation
+  - Docs summarise how the code should be used
+  - Docs reference specifications or design documents
+  - New public code has doc comments: `#![deny(missing_docs)]`
+  - Complex private code has doc comments
+
+- [ ] Tests
+  - The tests make sure the code implements the Zcash consensus rules
+  - Consensus rules have success and failure tests
```
Comment (teor2345) [inline .github/pull_request_template.md]:
```suggestion
  - Consensus rules have success tests, failure tests, and property tests
```

**Review by oxarbitrage — DISMISSED** [review]
I am ok in general as the reviewer guidelines are part of a flexible checklist and they are suggestions we will try to follow, not rules to apply to every PR.

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -26,3 +25,44 @@ the code change.
...
+
+Reviewer:
+This checklist can help you do your review.
+Add or skip tasks as needed.
+-->
+
+**Does this pull request improve Zebra?**
+
+- [ ] Pull Request
+  - The pull request is complete: code, documentation, and tests
```
Comment (teor2345) [inline .github/pull_request_template.md]:
```suggestion
  - **Is the pull request complete: code, documentation, and tests?**
```

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -26,3 +25,44 @@ the code change.
...
+  - The PR contains changes to related code (unrelated changes can be split into another PR)
+  - The PR does everything listed in the tickets that it closes
+  - Any follow-up tasks are listed in a GitHub issue
+
+- [ ] Code
+  - The code implements the specifications or design documents
```
Comment (teor2345) [inline .github/pull_request_template.md]:
```suggestion
  - **Does the code implement the specifications or design documents?**
```

File: .github/pull_request_template.md [hunk 5]
```diff
@@ -26,3 +25,44 @@ the code change.
...
+  - Changes from the specifications or design documents are explained in comments
+  - The code is readable, and does not appear to have any bugs
+  - Any known issues or limitations are documented
+
+- [ ] Documentation
+  - Docs summarise how the code should be used
```
Comment (teor2345) [inline .github/pull_request_template.md]:
```suggestion
  - **Do the docs summarise how the code should be used?**
```

File: .github/pull_request_template.md [hunk 6]
```diff
@@ -26,3 +25,44 @@ the code change.
...
+  - Docs reference specifications or design documents
+  - New public code has doc comments: `#![deny(missing_docs)]`
+  - Complex private code has doc comments
+
+- [ ] Tests
+  - The tests make sure the code implements the Zcash consensus rules
```
Comment (teor2345) [inline .github/pull_request_template.md]:
```suggestion
  - **Do the tests make sure the code implements the Zcash consensus rules?**
```

**Comment by teor2345** [comment]
I added design RFC questions, and follow-up work sections, so I've reset everyone's approvals.

File: .github/pull_request_template.md [hunk 7]
```diff
@@ -22,7 +23,55 @@ Summarize the solution and provide any necessary context needed to understand
...
+Add extra tasks to the review using list items.
+Skip review tasks using ~~strikethrough~~, or just delete them.
+If you want this pull request to have a specific reviewer, tag them in the list of reviewers.
+
+Reviewer:
+This checklist can help you do your review.
+Add or skip tasks as needed.
+-->
+
+**Does this pull request improve Zebra?**
```
Comment (yaahc) [inline .github/pull_request_template.md:52]: I think this should be removed, it reads as confrontational to me. I would expect every PR to improve zebra, otherwise it wouldn't have been written, and focusing on this question in the review checklist could be intimidating for new contributors, as though doubts were being cast upon their ability to improve and contribute to the codebase.

```suggestion
```

File: .github/pull_request_template.md [hunk 8]
```diff
@@ -22,7 +23,55 @@ Summarize the solution and provide any necessary context needed to understand
...
+- [ ] Pull Request
+  - **Is the pull request complete: code, documentation, and tests?**
+  - The PR contains changes to related code (unrelated changes can be split into another PR)
+  - The PR does everything listed in the tickets that it closes, or the designs that it finishes
+  - Any follow-up tasks are listed in a GitHub issue
+
+- [ ] Code
+  - **Does the code implement the specifications or design documents?**
+  - Changes from the specifications or design documents are explained in comments
+  - The code is readable, and does not appear to have any bugs
```
Comment (yaahc) [inline .github/pull_request_template.md]: The second item seems a little redundant.

```suggestion
  - The code prioritizes readability.
```

**Review by yaahc — COMMENTED** [review]
looks good, just a couple of suggestions that are mostly tone oriented

File: .github/pull_request_template.md (hunk 7 above)
Comment (teor2345) [reply .github/pull_request_template.md:52]: The goals of a PR and review are not stated, so authors and different reviewers can have conflicting goals - we've had PRs where that's happened.

For example, here is a non-goal:
Have we done everything possible to improve the code? Is it perfect?

And here are some unclear sub-goals:
When is documentation required?
When are tests required?

I'll try to rephrase the goal to be clearer.

Comment (teor2345) [reply .github/pull_request_template.md:52]:
```suggestion
**Overall Goal: Merged pull requests should improve Zebra.**
<!--
PRs should make incremental improvements to Zebra.
Reviewers can help PR authors write better code.
But PRs don't need to be perfect. It's ok to have follow-up tasks.
-->
```

Comment (yaahc) [reply .github/pull_request_template.md:52]: I think this still has those same unintentional connotation issues. How do you feel about this version?

```
- [ ] Description
  - **Does this PR focus on and achieve a clearly defined goal.**
  - The PR description highlights a concrete problem.
  - The solution directly solves the stated problem, rather than circumventing it (as possible).
  - The implemented code is not intended to be thrown away at a later date.
```

File: .github/pull_request_template.md (hunk 8 above)
Comment (teor2345) [reply .github/pull_request_template.md]: What about...
```suggestion
  - The code is readable and maintainable
```

File: .github/pull_request_template.md [hunk 9]
```diff
@@ -22,7 +23,55 @@ Summarize the solution and provide any necessary context needed to understand
...
+Reviewer:
+This checklist can help you do your review.
+Add or skip tasks as needed.
+-->
+
+**Does this pull request improve Zebra?**
+
+- [ ] Pull Request
+  - **Is the pull request complete: code, documentation, and tests?**
+  - The PR contains changes to related code (unrelated changes can be split into another PR)
```
Comment (teor2345) [inline .github/pull_request_template.md:56]:
```suggestion
  - The PR description identifies a specific problem and solution
  - The PR contains changes to related code (split unrelated changes into another PR)
```

File: .github/pull_request_template.md (hunk 7 above)
Comment (teor2345) [reply .github/pull_request_template.md:52]: I'd like to discuss the goals of a PR, because it looks like we have some different goals here.

>  - **Does this PR focus on and achieve a clearly defined goal.**

This is a useful goal, but it overlaps with a later goal:
The PR contains changes to related code (unrelated changes can be split into another PR)

>  - The PR description highlights a concrete problem.

I think we should add this goal to the pull request section:
https://github.com/ZcashFoundation/zebra/pull/1177/files#r513157688

>  - The solution directly solves the stated problem, rather than circumventing it (as possible).

This is a good goal, but I don't know how to check for it.
If we can find a good way to word it, let's add it to the Code section:
https://github.com/ZcashFoundation/zebra/pull/1177/files#diff-b2496e80299b8c3150b1944450bd81c622e04e13d15c411d291db0927d75fd6bR60

I also think there are good engineering reasons for circumventing some problems - as long as that's documented. See these other items:
- Any known issues or limitations are documented
- Any follow-up tasks are listed in a GitHub issue

>  - The implemented code is not intended to be thrown away at a later date.

I think this is a non-goal: it's ok to add code that we plan to throw away, as long as that's clearly documented using comments and issues. See these related items:
- The PR does everything listed in the tickets that it closes, or the designs that it finishes
- Any follow-up tasks are listed in a GitHub issue
- Any known issues or limitations are documented

-----

I am also trying to keep the template short.

So I'd like to limit the number of sections and checklist items. Let's also look for things to delete - we should delete anything that isn't a top priority. (We can move details to the Zebra book if we need to.)

File: .github/pull_request_template.md [hunk 10]
```diff
@@ -22,7 +23,55 @@ Summarize the solution and provide any necessary context needed to understand
...
+  - Any follow-up tasks are listed in a GitHub issue
+
+- [ ] Code
+  - **Does the code implement the specifications or design documents?**
+  - Changes from the specifications or design documents are explained in comments
+  - The code is readable, and does not appear to have any bugs
+  - Any known issues or limitations are documented
+
+- [ ] Documentation
+  - **Do the docs summarise how the code should be used?**
```
Comment (teor2345) [inline .github/pull_request_template.md:67]: Should we delete this checklist item?
```suggestion
  - **Docs summarise what the code does, and how it should be used**
```

File: .github/pull_request_template.md [hunk 11]
```diff
@@ -22,7 +23,55 @@ Summarize the solution and provide any necessary context needed to understand
...
+  - Docs reference specifications or design documents
+  - New public code has doc comments: `#![deny(missing_docs)]`
+  - Complex private code has doc comments
+
+- [ ] Tests
+  - **Do the tests make sure the code implements the Zcash consensus rules?**
+  - Consensus rules have success tests, failure tests, and property tests
+  - Edge cases and errors have tests
```
Comment (teor2345) [inline .github/pull_request_template.md:75]: Should we delete this checklist item?
```suggestion

```

File: .github/pull_request_template.md [hunk 12]
```diff
@@ -22,7 +23,55 @@ Summarize the solution and provide any necessary context needed to understand
...
+- [ ] Code
+  - **Does the code implement the specifications or design documents?**
+  - Changes from the specifications or design documents are explained in comments
+  - The code is readable, and does not appear to have any bugs
+  - Any known issues or limitations are documented
+
+- [ ] Documentation
+  - **Do the docs summarise how the code should be used?**
+  - Docs reference specifications or design documents
+  - New public code has doc comments: `#![deny(missing_docs)]`
```
Comment (teor2345) [inline .github/pull_request_template.md:69]: I want to add `#![deny(missing_docs)]` to every zebra crate (#453), then we can remove this checklist item.

File: .github/pull_request_template.md (hunk 7 above)
Comment (teor2345) [reply .github/pull_request_template.md:52]: Should we delete this checklist item?

The item is ambiguous - it means different things to different people. So a PR template isn't a good place to discuss the goals of a review.

Let's move this discussion to next week's Zebra sync?

```suggestion
```

Comment (yaahc) [reply .github/pull_request_template.md:52]: :+1:

**Comment by teor2345** [comment]
I think this version of the template is far too long.
Reviewer guidelines belong in the book, not a template.

See #1235 for a much shorter version.
